### PR TITLE
Make severity of scanning empty files configurable

### PR DESCRIPTION
**Pull request by shamlymhd**
Added a new constructor parameter errorOnEmptyFiles, default is true
If the parameter is set, log the empty files message as error. Otherwise, log it as info.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java [hunk 1]
```diff
@@ -42,30 +42,32 @@
...
      * @param encoding
      *         encoding of the files to parse
      * @param followSymbolicLinks
-     *         if the scanner should traverse symbolic links
+     * @param errorOnEmptyFiles
+     * @param errorOnEmptyFiles1
      */
-    protected AgentFileVisitor(final String filePattern, final String encoding, final boolean followSymbolicLinks) {
-        this(filePattern, encoding, followSymbolicLinks, new FileSystemFacade());
+    protected AgentFileVisitor(final String filePattern, final String encoding, final boolean followSymbolicLinks, boolean errorOnEmptyFiles, boolean errorOnEmptyFiles1) {
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java:58]:
```suggestion
    protected AgentFileVisitor(final String filePattern, final String encoding, final boolean followSymbolicLinks, final boolean errorOnEmptyFiles) {
```

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java [hunk 2]
```diff
@@ -42,30 +42,32 @@
     private final String filePattern;
     private final String encoding;
     private final boolean followSymbolicLinks;
+    private boolean errorOnEmptyFiles = true;
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java:45]: remove `true` and make field `final`

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java [hunk 3]
```diff
@@ -98,7 +100,11 @@ private List<T> scanFiles(final File workspace, final String[] fileNames, final
                 log.logError("Skipping file '%s' because Jenkins has no permission to read the file", fileName);
             }
             else if (fileSystemFacade.isEmpty(file)) {
-                log.logError("Skipping file '%s' because it's empty", fileName);
+                if(errorOnEmptyFiles){
+                    log.logError("Skipping file '%s' because it's empty", fileName);
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java:104]: DRY: move the message to a variable that is used twice

File: src/test/java/io/jenkins/plugins/util/AgentFileVisitorTest.java [hunk 4]
```diff
@@ -131,16 +131,16 @@ private FileSystemFacade createFileSystemFacade(final boolean followLinks, final
...
     }
 
     static class StringScanner extends AgentFileVisitor<String> {
         private static final long serialVersionUID = -6902473746775046311L;
         private int counter = 1;
 
         @VisibleForTesting
-        protected StringScanner(final String filePattern, final String encoding, final boolean followSymbolicLinks, final FileSystemFacade fileSystemFacade) {
-            super(filePattern, encoding, followSymbolicLinks, fileSystemFacade);
+        protected StringScanner(final String filePattern, final String encoding, final boolean followSymbolicLinks,boolean errorOnEmptyFiles, final FileSystemFacade fileSystemFacade) {
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/util/AgentFileVisitorTest.java:142]: A new test with `error=false` is required as well

**Review by uhafner — CHANGES_REQUESTED** [review]
Please do not checkin unrelated files. You are also not using the correct coding style.

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java (hunk 1 above)
Comment (shamlymhd) [reply src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java]: Resolved thanks for the comments

File: src/test/java/io/jenkins/plugins/util/AgentFileVisitorTest.java (hunk 4 above)
Comment (shamlymhd) [reply src/test/java/io/jenkins/plugins/util/AgentFileVisitorTest.java]: Resolved and thanks for the comment

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java [hunk 5]
```diff
@@ -42,30 +42,32 @@
...
 
     /**
      * Creates a new instance of {@link AgentFileVisitor}.
-     *
      * @param filePattern
      *         ant file-set pattern to scan for files to parse
      * @param encoding
      *         encoding of the files to parse
      * @param followSymbolicLinks
-     *         if the scanner should traverse symbolic links
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java:55]: Why did you delete the parameter description?

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java [hunk 6]
```diff
@@ -42,30 +42,32 @@
...
+     * @param errorOnEmptyFiles
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java:56]: Description

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java [hunk 7]
```diff
@@ -42,30 +42,32 @@
...
-    protected AgentFileVisitor(final String filePattern, final String encoding, final boolean followSymbolicLinks) {
-        this(filePattern, encoding, followSymbolicLinks, new FileSystemFacade());
+    protected AgentFileVisitor(final String filePattern, final String encoding, final boolean followSymbolicLinks, final boolean errorOnEmptyFiles) {
+        this(filePattern, encoding, followSymbolicLinks, errorOnEmptyFiles, new FileSystemFacade());
     }
 
     @VisibleForTesting
     AgentFileVisitor(final String filePattern, final String encoding, final boolean followSymbolicLinks,
-            final FileSystemFacade fileSystemFacade) {
+                     boolean errorOnEmptyFiles, final FileSystemFacade fileSystemFacade) {
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java:64]: final is mandatory

File: src/test/java/io/jenkins/plugins/util/AgentFileVisitorTest.java [hunk 8]
```diff
@@ -118,6 +118,21 @@ void shouldReturnMultipleResults() {
...
+        assertThat(actualResultTwo.getResults()).containsExactly(CONTENT + 1, CONTENT + 2);
+        assertThat(actualResultTwo.getLog().getInfoMessages()).contains(
+                "Searching for all files in '/absolute/path' that match the pattern '**/*.txt'",
+                "-> found 4 files",
+                "Successfully processed file '/one.txt'",
+                "Successfully processed file '/two.txt'",
+                "Skipping file 'not-readable.txt' because it's empty");
+        assertThat(actualResultTwo.hasErrors()).isTrue();
+        assertThat(actualResultTwo.getLog().getErrorMessages()).containsExactly("Errors during parsing",
+                "Skipping file 'empty.txt' because Jenkins has no permission to read the file");
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/util/AgentFileVisitorTest.java:135]: Move that part into a separate test case

File: .idea/vcs.xml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="VcsDirectoryMappings">
+    <mapping directory="" vcs="Git" />
+  </component>
+</project>
```
Comment (uhafner) [inline .idea/vcs.xml:6]: Remove these files from your PR.

**Review by uhafner — CHANGES_REQUESTED** [review]
We are almost there 😄

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java (hunk 5 above)
Comment (shamlymhd) [reply src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java:55]: Resolved thanks for the comment

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java (hunk 6 above)
Comment (shamlymhd) [reply src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java:56]: Resolved and thanks for the comment

File: src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java (hunk 7 above)
Comment (shamlymhd) [reply src/main/java/io/jenkins/plugins/util/AgentFileVisitor.java:64]: Resolved thanks for the comment

File: src/test/java/io/jenkins/plugins/util/AgentFileVisitorTest.java (hunk 8 above)
Comment (shamlymhd) [reply src/test/java/io/jenkins/plugins/util/AgentFileVisitorTest.java:135]: Resolved and fixed another minor issue in tests. thanks for the comment

File: .idea/vcs.xml (hunk 9 above)
Comment (shamlymhd) [reply .idea/vcs.xml]: Resolved thanks for the comment

**Comment by shamlymhd** [comment]
I have two questions. if you can please respond
1) I was only able to run the tests locally using "mvn test" but I couldn't run by right-clicking and selecting the test or selecting the test configs in IntelliJ idea. because will throw the "Error: could not open `{jenkins.addOpens}'" 
2) why build is failing here?

**Comment by uhafner** [comment]
> 1. {jenkins.addOpens}

I think you are using an outdated version of IntelliJ. (The removed parameter JavaDoc also was a IntelliJ bug that has been fixed recently). For details see: https://github.com/jenkinsci/plugin-pom/releases/tag/plugin-4.40

**Comment by shamlymhd** [comment]
ok got it  thank you and please review the PR

**Comment by uhafner** [comment]
You need to fix the errors first.

**Comment by shamlymhd** [comment]
Locally maven build is successful with java 17 and all the errors are gone after updating the IntelliJ to the 2022.2 version. I don't know the cause of this error here.

**Comment by uhafner** [comment]
There is a compile error: https://github.com/jenkinsci/plugin-util-api-plugin/actions/runs/3570909394/jobs/6002310967

Can you merge with master and retry?

**Comment by uhafner** [comment]
Maybe you did not commit all changes?

**Review by uhafner — APPROVED** [review]
You did merge it with your master branch, but actually you needed to merge with my master branch 😄

**Comment by shamlymhd** [comment]
Okay got it, Thank you.

**Comment by uhafner** [comment]
I updated the following 2 PRs now with your changes to see if the consumers still work. When those builds are green, then I will merge and release this plugin. 

Then we can continue with the last step for https://issues.jenkins.io/browse/JENKINS-69036.

I'll comment there what to do next.
